proposals/005_New_Security_Definitions.md

```diff
@@ -0,0 +1,531 @@
+# New Security Definitions
+
+## Metadata
+
+| Tag | Value |
+| ---- | ---------------- |
+| Proposal | [005](https://github.com/OAI/OpenAPI-Specification/tree/main/proposals/005_New_Security_Definitions.md) |
+| Authors | [Jeremy Whitlock](https://github.com/whitlockjc) |
+| Review Manager | TBD |
+| Status | Proposal |
+| Implementations | N/A |
+| Issues | [{issueid}](https://github.com/OAI/OpenAPI-Specification/issues/{Issueid}) |
+| Previous Revisions | N/A |
+
+## Change Log
+
+| Date | Responsible Party | Description |
+| 2021/05/13 | Jeremy Whitlock | Initial proposal |
+
+## Introduction
+
+This purpose of this proposal is to rethink how
+[Security Schemes](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#security-scheme-object) are
+defined in OpenAPI.  This proposal would separate the credential location(s) from the Security Scheme _type_ and would
+allow for complete control over dictating where the credential(s) are provided for a Security Scheme.  This proposal
+will add a generic way to retrieve a credential from anywhere in a request, and allow any Security Scheme to dictate
+where its supported credential(s) are instead of relying on OpenAPI-supported conventions/opinions.  This proposal also
+separates the credential location(s) details from the other configurations associated with the Security Scheme
+_(like JWKS details for JWT)_.
+
+**Note:** The proposal isn't to remove `type` from the Security Scheme object but to allow for more control over where
+the credential(s) are provided for a Security Scheme.  This proposal would still have reasonable defaults for the
+already-supported `type` values and will be documented as such.
+
+## Motivation
+
+[Security Schemes](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#security-scheme-object)
+_(formerly known as "Security Definitions")_ provide a way to describe the security scheme(s) used by your API.
+Unfortunately, the _types_ supported by OpenAPI are not only limited but they often tie the credential location to the
```

Comment:
> Unfortunately, the _types_ supported by OpenAPI ... often tie the credential location to the Security Scheme type

Can you expand on why this is "unfortunate"? Type `apiKey` can already be located `in` any of `query`, `header`,`cookie`, and arguably if you try to put your `Bearer` token anywhere other than in an `Authorization` header, you **are not doing oAuth2**.

Reply:
The point was more about separating the credential location(s) from the Security Scheme type.  API Key is the closest to allowing free-form credential locations while others do not.  I can see where some Security Schemes where these are locked down, like OAuth 2.0 _(while it uses Bearer Tokens, I don't think it supports them being in all locations a Bearer Token could be represented)_, but this separation is really a flexibility thing than just trying to support JWT.

Reply:
It would also be nice to have a way to describe a type not supported by OAS, where our locked-down types don't fit and a customer wants to still describe their security needs.